core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -669,6 +686,34 @@ private[spark] class AppStatusListener(
         }
       }
     }
+
+    // check if there is a new peak value for any of the executor level memory metrics
+    // for the live UI. SparkListenerExecutorMetricsUpdate events are only processed
+    // for the live UI.
+    event.executorUpdates.foreach { updates: ExecutorMetrics =>
+      liveExecutors.get(event.execId).foreach { exec: LiveExecutor =>
+        if (exec.peakExecutorMetrics.compareAndUpdatePeakValues(updates)) {
+          maybeUpdate(exec, now)
+        }
+      }
+    }
+  }
+
+  override def onStageExecutorMetrics(executorMetrics: SparkListenerStageExecutorMetrics): Unit = {
+    val now = System.nanoTime()
+
+    // check if there is a new peak value for any of the executor level memory metrics,
+    // while reading from the log. SparkListenerStageExecutorMetrics are only processed
+    // when reading logs.
+    liveExecutors.get(executorMetrics.execId)
+      .orElse(deadExecutors.get(executorMetrics.execId)) match {
+      case Some(exec) =>
```

Comment:
Don't case compare with `Some` and `None` - for options prefer the functional equivalents and occasionally can use isEmpty / isPresent. We can do this here:

```
val opt = liveExecutors.get(...).orElse(...)
opt.foreach { ...}
if (opt.isEmpty) {
  logWarning(...)
}
```

Reply:
I don't think we really make a point about this code style.  In general we prefer the functional versions, but I actually find a match on Some & None totally fine.  There are plenty of other examples of this in the code base (though that doesn't necessarily mean its the right style ...)

Reply:
yeah, but you're talking about both a `foreach` *and* an `if` together.

A long time back we discussed using `option.fold` for this, as it is all in one function, but we rejected it as being pretty confusing for most developers.

```scala
scala> def foo(x: Option[String]) = x.fold("nada")("some " + _)
foo: (x: Option[String])String

scala> foo(None)
res0: String = nada

scala> foo(Some("blah"))
res1: String = some blah
```

Reply:
Ok, this can stay as is.